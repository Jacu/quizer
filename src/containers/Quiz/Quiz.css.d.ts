export const Quiz: string;
export const Questions: string;
