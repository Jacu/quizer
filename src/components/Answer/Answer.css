.Answer {
    display: inline-block;
    width: 50%;        
    /* border: 1px solid rgba(0,0,0,0.1); */
    padding: 15px;
    cursor: pointer;
}

.selected {
    background: #125271;
    color: white; 
}

.correct {
    background: green;
    color: white;
}

.incorrect {
    background: red;
    color: white;
}

.Answer:hover {
    background: #b3d8e9; 
}