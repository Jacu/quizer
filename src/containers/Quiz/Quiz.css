.Quiz {
    background: white;
    position: relative;    
    align-self: center;
    width: 60%;    
    height: 50%;
    border-radius: 50px;
    box-shadow: 2px 2px 10px 2px rgba(0,0,0,0.25);
    display: flex; 
    justify-content: center;
    align-items: center;
}

.arrow {
    padding: 20px;
    flex: 1;
    text-decoration: none;
    color: inherit;
    font-weight: bold;
    font-size: 5rem;
}