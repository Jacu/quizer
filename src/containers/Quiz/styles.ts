import styled from "styled-components";

interface IQuizStyle {
    finished: boolean, 
    ref: React.MutableRefObject<HTMLDivElement | null>,
}

export const Quiz = styled.div<IQuizStyle>`
    align-self: center;
    width: 80%;
    max-width: 700px;
    height: 100%;
    display: flex;
    justify-content: ${props => props.finished ? 'start' : 'center'};
    align-items: center;
    flex-direction: column;
    :focus {
        outline: none;
    }
`;

export const ButtonContainer = styled.div`
    padding-bottom: 60px;
`;