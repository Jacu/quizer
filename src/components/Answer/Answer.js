import React from 'react';
import styles from './Answer.css'

const Answer = ({selected,onClick,answer,correct,ended}) => {    
    const className = [styles.Answer,selected ? ended ? correct ? styles.correct : styles.incorrect : styles.selected : ended ? correct ? styles.correct : '' : ''];
    return (
    <div 
        className={className.join(' ')}
        onClick={ended ? null : onClick}>
        {answer}
    </div>)
};

export default Answer;