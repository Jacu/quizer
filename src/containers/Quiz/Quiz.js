import React from 'react';
import styles from './Quiz.css';
import Question from '../../components/Question/Question';
import { connect } from 'react-redux';
import { Route, Redirect } from 'react-router-dom';
import * as actions from '../../store/actions/index';
import Spinner from '../../components/UI/Spinner/Spinner';
import Arrow from '../../components/UI/Arrow/Arrow';
import Button from '../../components/UI/Button/Button';

const Quiz = (props) => {    
    const loading = !(props.quizStarted || props.quizFinished) || props.dataLoading;

    return (
        <div className={styles.Quiz}>
            <div className={styles.Questions}>
                <Arrow
                    disable={props.currentQuestion <= 1}
                    direction="left">
                </Arrow>
                {/* {!loading && !props.quizStarted ? <Redirect to="/" /> : null} */}
                {loading ? <Spinner /> : <Route path="/quiz/:id" component={Question} /> }
                <Arrow
                    disable={props.currentQuestion >= props.questionsAmount}>
                </Arrow>
            </div>
            <Button link='/summary' label={props.ended ? "View score" : "Submit"} onClick={!props.ended ? props.end : null} />
        </div>
    )
}

const mapStateToProps = ({quiz,startPage}) => {
    return {
        currentQuestion: quiz.questions.current,
        questionsAmount: quiz.questions.amount,
        quizStarted: quiz.started,
        quizFinished: quiz.finished,        
        ended: quiz.finished,
        dataLoading: quiz.questions.fetching,
    }
}

const mapDispatchToProps = dispatch => {
    return {        
        end: () => dispatch(actions.quizEnded())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Quiz);