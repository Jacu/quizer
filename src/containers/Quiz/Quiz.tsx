import React, { useState } from 'react';
import * as styled from './styles';
import Question from '../../components/Question/Question';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import * as actions from '../../store/actions/index';
import Spinner from '../../components/UI/Spinner/Spinner';
import NavBar from "../../components/NavBar/NavBar";
import Button from '../../components/UI/Button/Button';
import { AppState } from '../../index';
import Summary from "~/components/Question/Summary/Summary";
import { Question as IQuiestion } from "~/store/reducers/quiz";

interface QuizProps {
}

interface StateProps {
    questions: IQuiestion[],
    quizStarted: boolean,
    dataLoading: boolean,
    isDataAvailable: boolean,
}

interface DispatchProps {
    end: () => void,
    initQuiz: () => void,
}

type Props = QuizProps & StateProps & DispatchProps;

const Quiz: React.FC<Props> = props => {
    const [finished, setFinished] = useState(false);
    const [questionNumber, setQuestionNumber] = useState(1);
    const [correctAnswers, setCorrectAnswers] = useState<boolean[]>([]);
    const questionId = questionNumber - 1;
    const [selectedId, setSelectedId] = useState(-1);
    const question = props.questions[questionId];

    let answers: string[] = [];
    const questionsAmount = props.questions.length;
    const correctAnswerId = question?.correctAnswerId;
    const score = correctAnswers.reduce((prev,curr) => curr ? prev + 1 : prev, 0) / questionsAmount * 100;

    const [inRevealMode, setInRevealMode] = useState(false);
    const buttonLabel = finished ? 'Try Again!' : inRevealMode ? "Next" : "Check";
    const loading = !(props.quizStarted || finished) || props.dataLoading;

    if(!props.isDataAvailable) {
        return <Redirect to="/" />
    };

    if (question) {
        answers = [...question.incorrect_answers];
        answers.splice(correctAnswerId, 0, question.correct_answer);
    }

    const resetSettings = () => {
        setQuestionNumber(1);
        setCorrectAnswers([]);
        setFinished(false);
        setInRevealMode(false)
        props.initQuiz();
    }

    const handleButtonClick = () => {
        if (finished) {
            resetSettings();
        } else if (inRevealMode) {
            const isLastQuestion = questionNumber === props.questions.length;
            setSelectedId(-1);
            setFinished(isLastQuestion);
            setQuestionNumber(isLastQuestion ? questionNumber : questionNumber  + 1);
            setInRevealMode(false);
        } else {
            const isCorrect = selectedId === correctAnswerId
            setCorrectAnswers([...correctAnswers, isCorrect])
            setInRevealMode(true);
        };
    }

    const handleAnswerPicked = (selectedAnswerId: number) => {
        setSelectedId(selectedAnswerId);
    }

    return (
        <styled.Quiz finished={finished}>
            {loading ? <Spinner /> :
                !finished
                    ? <Question
                        question={props.questions[questionId]}
                        reveal={inRevealMode}
                        answers={answers}
                        correctId={correctAnswerId}
                        onAnswerPicked={handleAnswerPicked}
                        selectedId={selectedId} />
                    : <Summary score={score} questions={props.questions} correctAnswers={correctAnswers}  />}
            <Button label={buttonLabel} onClick={handleButtonClick} />
            <NavBar 
                currentQuestionNumber={questionNumber}
                questionsAmount={questionsAmount}
                loading={props.dataLoading}
                finished={finished} />
        </styled.Quiz>
    )
}

const mapStateToProps = ({ quiz, startPage }: AppState): StateProps => {
    return {
        questions: quiz.questions,
        quizStarted: quiz.started,
        dataLoading: quiz.fetching,
        isDataAvailable: startPage.dataFetched,
    }
}

const mapDispatchToProps = (dispatch): DispatchProps => {
    return {
        end: () => dispatch(actions.quizEnded()),
        initQuiz: () => dispatch(actions.initQuiz()),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Quiz);