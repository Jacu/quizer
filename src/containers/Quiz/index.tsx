import React, { useState, useRef, useEffect } from 'react';
import * as styled from './styles';
import Question from '../../components/Question';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import * as actions from '../../store/actions/index';
import { QuizActions } from '../../store/actions/actionTypes';
import Spinner from '../../components/UI/Spinner';
import NavBar from "~/components/NavBar";
import Button from '../../components/UI/Button';
import { AppState } from '../../index';
import Summary from "~/components/Summary";
import { IQuestion } from "~/store/reducers/quiz";
import KEYS from "~/utils/keys";

interface StateProps {
    questions: IQuestion[],
    quizStarted: boolean,
    dataLoading: boolean,
    isDataAvailable: boolean,
}

interface DispatchProps {
    end: () => void,
    initQuiz: () => void,
    quit: () => void,
}

const Quiz: React.FC<StateProps & DispatchProps> = props => {
    const [finished, setFinished] = useState(false);
    const [questionNumber, setQuestionNumber] = useState(1);
    const [correctAnswers, setCorrectAnswers] = useState<boolean[]>([]);
    const questionId = questionNumber - 1;
    const [selectedId, setSelectedId] = useState<number>();
    const question = props.questions[questionId];
    const quizRef = useRef<HTMLDivElement>(null);

    let answers: string[] = [];
    const questionsAmount = props.questions.length;
    const correctAnswerId = question?.correctAnswerId;
    const score = correctAnswers.reduce((prev, curr) => curr ? prev + 1 : prev, 0) / questionsAmount * 100;

    const [inRevealMode, setInRevealMode] = useState(false);
    const buttonLabel = finished ? 'Try Again!' : inRevealMode ? "Next" : "Check";
    const loading = !(props.quizStarted || finished) || props.dataLoading;

    useEffect(() => {
        if (quizRef?.current) {
            quizRef.current!.focus();
        }
    }, [quizRef]);

    if (!props.isDataAvailable) {
        return <Redirect to="/" />
    };

    if (question) {
        answers = [...question.incorrect_answers];
        answers.splice(correctAnswerId, 0, question.correct_answer);
    }

    const resetSettings = () => {
        setQuestionNumber(1);
        setCorrectAnswers([]);
        setFinished(false);
        setInRevealMode(false)
        props.initQuiz();
    }

    const handleButtonClick = () => {
        if (finished) {
            resetSettings();
        } else if (inRevealMode) {
            const isLastQuestion = questionNumber === props.questions.length;
            setSelectedId(undefined);
            setFinished(isLastQuestion);
            setQuestionNumber(isLastQuestion ? questionNumber : questionNumber + 1);
            setInRevealMode(false);
        } else {
            const isCorrect = selectedId === correctAnswerId
            setCorrectAnswers([...correctAnswers, isCorrect])
            setInRevealMode(true);
        };
    }

    const handleAnswerPicked = (selectedAnswerId: number) => {
        setSelectedId(selectedAnswerId);
    }

    const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
        const isBooleanQuestion = answers.length <= 2;
        console.log(e.key);
        
        if (!loading) {
            switch (e.key) {
                case KEYS.ENTER:
                    handleButtonClick();
                    break;
                case KEYS.ESCAPE:
                    props.quit();
                    break;
                case KEYS.ARROW_UP:
                    if (!inRevealMode && selectedId != null && selectedId > 0) {
                        setSelectedId(selectedId - 1);
                        break;
                    }
                    !inRevealMode &&setSelectedId(isBooleanQuestion ? 1 : 3);
                    break;
                case KEYS.ARROW_DOWN:
                    if (!inRevealMode &&selectedId != null && (isBooleanQuestion ? selectedId < 1 : selectedId < 3)) {
                        setSelectedId(selectedId + 1);
                        break;
                    }
                    !inRevealMode && setSelectedId(0);
                    break;
                case KEYS.ONE:
                    !inRevealMode && setSelectedId(0);
                    break;
                case KEYS.TWO:
                    !inRevealMode && setSelectedId(1);
                    break;
                case KEYS.THREE:
                    !isBooleanQuestion && !inRevealMode && setSelectedId(2);
                    break;
                case KEYS.FOUR:
                    !isBooleanQuestion && !inRevealMode && setSelectedId(3);
                    break;
                default:
                    break;
            }
        }
    };

    const getContent = () => {
        if (loading) {
            return <Spinner />
        } else if (finished) {
            return <Summary 
                score={score} 
                questions={props.questions.map((q,i) => ({question: q.question, answer: q.correct_answer, isCorrect: correctAnswers[i]}))} />
        }
        const currentQuestion = props.questions[questionId];
        return <Question
            question={currentQuestion.question}
            category={currentQuestion.category}
            reveal={inRevealMode}
            answers={answers}
            correctId={correctAnswerId}
            onAnswerPicked={handleAnswerPicked}
            selectedId={selectedId} />
    }

    return (
        <styled.Quiz finished={finished} tabIndex={0} onKeyDown={handleKeyDown} ref={quizRef}>
            {getContent()}
            <Button label={buttonLabel} onClick={handleButtonClick} />
            <NavBar
                currentQuestionNumber={questionNumber}
                questionsAmount={questionsAmount}
                loading={props.dataLoading}
                onQuit={props.quit}
                finished={finished} />
        </styled.Quiz>
    )
}

const mapStateToProps = ({ quiz, startPage }: AppState): StateProps => {
    return {
        questions: quiz.questions,
        quizStarted: quiz.started,
        dataLoading: quiz.fetching,
        isDataAvailable: startPage.dataFetched,
    }
}

const mapDispatchToProps = (dispatch: ThunkDispatch<AppState, null, QuizActions>): DispatchProps => {
    return {
        end: () => dispatch(actions.endQuiz()),
        initQuiz: () => dispatch(actions.initQuiz()),
        quit: () => dispatch(actions.quizQuit()),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Quiz);