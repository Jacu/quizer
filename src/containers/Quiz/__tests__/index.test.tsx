import React from 'react';
import Quiz from "../";
import { render, fireEvent, waitFor, waitForElementToBeRemoved, screen, getByText } from "@testing-library/react";
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { initQuiz, quizQuit } from "../../../store/actions";
import { AnswerStates } from "~/components/Question/Answer/config";
import '@testing-library/jest-dom';

const BUTTON_LABEL = {
    DEFAULT: 'Check',
    IN_REVEAL: 'Next',
    RESULTS: 'Try Again!'
};

const mockQuestion1 = {
    category: 'mock',
    type: 'obvious',
    difficulty: 'no brainer',
    question: 'Is test passing?',
    correct_answer: 'of course!',
    incorrect_answers: ['no', 'not yet'],
    correctAnswerId: 1,
};

const mockQuestion2 = {
    category: 'not mock',
    type: 'complex',
    difficulty: 'tricky',
    question: 'What?',
    correct_answer: 'Floccinaucinihilipilification',
    incorrect_answers: ['this', 'else'],
    correctAnswerId: 1,
};

const initialState = {
    startPage: {
        dataFetched: true,
    },
    quiz: {
        questions: [
            mockQuestion1,
            mockQuestion2,
        ],
        started: true,
        fetching: false,
    },
};

jest.mock('../../../store/actions', () => ({
    actionTypes: {},
    endQuiz: jest.fn(() => ({ type: 'type' })),
    initQuiz: jest.fn(() => ({ type: 'type' })),
    quizQuit: jest.fn(() => ({ type: 'type' })),
}));

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

describe('Quiz', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.clearAllTimers();
        jest.resetModules();
        store.clearActions();
    });

    it('should render with props', () => {
        const { asFragment } = render(<Provider store={store}><Quiz /></Provider>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should toggle button label', () => {
        const { getByText, queryByText } = render(<Provider store={store}><Quiz /></Provider>);
        expect(getByText(BUTTON_LABEL.DEFAULT)).toBeTruthy();
        expect(queryByText(BUTTON_LABEL.IN_REVEAL)).toBeNull();
        fireEvent.click(getByText(BUTTON_LABEL.DEFAULT));
        expect(queryByText(BUTTON_LABEL.DEFAULT)).toBeNull();
        expect(getByText(BUTTON_LABEL.IN_REVEAL)).toBeTruthy();
        fireEvent.click(getByText(BUTTON_LABEL.IN_REVEAL));
        expect(getByText(BUTTON_LABEL.DEFAULT)).toBeTruthy();
        expect(queryByText(BUTTON_LABEL.IN_REVEAL)).toBeNull();
    });

    it('should have all answers initialy unselected', () => {
        const numberOfCorrectAnswers = 1;
        const numberOfIncorrectAnswers = initialState.quiz.questions[0].incorrect_answers.length;
        const allAnswers = numberOfCorrectAnswers + numberOfIncorrectAnswers;
        const { getAllByTestId } = render(<Provider store={store}><Quiz /></Provider>);
        expect(getAllByTestId(AnswerStates.Notselected)).toHaveLength(allAnswers);
    });

    describe('should select correct answer on "Check" button click', () => {
        it('when nothing is selected', () => {
            const { getByText, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
            fireEvent.click(getByText('Check'));
            expect(getByTestId(AnswerStates.CorrectSelected)).toBeTruthy();
        });

        it('when incorrect answer is selected', () => {
            const incorrectAnswerLabel = initialState.quiz.questions[0].incorrect_answers[0];
            const { getByText, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
            fireEvent.click(getByText(incorrectAnswerLabel));
            fireEvent.click(getByText('Check'));
            expect(getByTestId(AnswerStates.CorrectSelected)).toBeTruthy();
            expect(getByTestId(AnswerStates.IncorrectSelected)).toBeTruthy();
        });

        it('when corect answer is selected', () => {
            const correctAnswerLabel = initialState.quiz.questions[0].correct_answer;
            const { getByText, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
            fireEvent.click(getByText(correctAnswerLabel));
            fireEvent.click(getByText('Check'));
            expect(getByTestId(AnswerStates.CorrectSelected)).toBeTruthy();
        });
    });

    it('should load next question on "Next" click', () => {
        const { getByText, queryByText } = render(<Provider store={store}><Quiz /></Provider>);
        const currentQuestion = initialState.quiz.questions[0].question;
        const nextQuestion = initialState.quiz.questions[1].question;
        expect(queryByText(currentQuestion)).toBeTruthy();
        expect(queryByText(nextQuestion)).toBeNull();
        fireEvent.click(getByText(BUTTON_LABEL.DEFAULT));
        fireEvent.click(getByText(BUTTON_LABEL.IN_REVEAL));
        expect(queryByText(currentQuestion)).toBeNull();
        expect(queryByText(nextQuestion)).toBeTruthy();
    });

    it('should render answers in correct order', () => {
        const { correctAnswerId, correct_answer } = initialState.quiz.questions[0];
        const { getAllByTestId } = render(<Provider store={store}><Quiz /></Provider>);
        expect(getAllByTestId(AnswerStates.Notselected)[correctAnswerId]).toHaveTextContent(correct_answer);
    });

    describe('should handle KeyDown', () => {
        describe('on Enter', () => {
            it('when not in reveal mode should set reveal mode', () => {
                const { container, getAllByText } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' });
                expect(getAllByText('Next')).toHaveLength(1);
            });

            it('when in reveal mode and not last question should load next question', () => {
                const currentQuestion = initialState.quiz.questions[0].question;
                const nextQuestion = initialState.quiz.questions[1].question;
                const { container, getByText, queryByText } = render(<Provider store={store}><Quiz /></Provider>);
                expect(queryByText(currentQuestion)).toBeTruthy();
                expect(queryByText(nextQuestion)).toBeNull();
                fireEvent.click(getByText(BUTTON_LABEL.DEFAULT));
                fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' });
                expect(queryByText(currentQuestion)).toBeNull();
                expect(queryByText(nextQuestion)).toBeTruthy();
            });

            it('after checking last question, should load Summary', () => {
                const { questions } = initialState.quiz;
                const { getByText, container } = render(<Provider store={store}><Quiz /></Provider>);
                const element = container.firstChild!;
                questions.forEach(q => {
                    fireEvent.keyDown(element, { key: 'Enter', code: 'Enter' }); // reveal correct answer
                    fireEvent.keyDown(element, { key: 'Enter', code: 'Enter' }); // load new question
                });
                expect(getByText(BUTTON_LABEL.RESULTS)).toBeTruthy();
            });

            it('should init new quiz when in summary view', () => {
                const { questions } = initialState.quiz;
                const { getByText, container } = render(<Provider store={store}><Quiz /></Provider>);
                questions.forEach(q => {
                    fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' }); // reveal correct answer
                    fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' }); // load new question
                });
                fireEvent.click(getByText(BUTTON_LABEL.RESULTS));
                expect(initQuiz).toHaveBeenCalled();
            });
        });
        describe('on ESCAPE', () => {
            it('should quitQuiz', () => {
                const { container } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key: 'Escape' });
                expect(quizQuit).toHaveBeenCalled();
            });
        });
        const { incorrect_answers, correct_answer, correctAnswerId } = initialState.quiz.questions[0];
        const allAnswers = [...incorrect_answers];
        allAnswers.splice(correctAnswerId, 0, correct_answer);
        describe('on ArrowUp', () => {
            it('when nothing is selected should select last answer', () => {
                const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key: 'ArrowUp', code: 'ArrowUp' });
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[allAnswers.length - 1]);
            });

            describe('when selection already present', () => {
                it('and not first', () => {
                    const selected = 1;
                    const shouldSelect = selected - 1;
                    const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                    fireEvent.click(getByText(container, allAnswers[selected]));
                    fireEvent.keyDown(container.firstChild!, { key: 'ArrowUp', code: 'ArrowUp' });
                    expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[shouldSelect]);
                });

                it('and is first', () => {
                    const selected = 0;
                    const shouldSelect = allAnswers.length - 1;
                    const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                    fireEvent.click(getByText(container, allAnswers[selected]));
                    fireEvent.keyDown(container.firstChild!, { key: 'ArrowUp', code: 'ArrowUp' });
                    expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[shouldSelect]);
                });
            });

            it('should not change selection in reveal mode', () => {
                const { container, getByTestId, queryAllByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.click(getByText(container, allAnswers[correctAnswerId]));
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[correctAnswerId]);
                fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' });
                fireEvent.keyDown(container.firstChild!, { key: 'ArrowUp', code: 'ArrowUp' });
                expect(getByTestId(AnswerStates.CorrectSelected)).toHaveTextContent(allAnswers[correctAnswerId]);
                expect(queryAllByTestId(AnswerStates.Notselected)).toHaveLength(allAnswers.length - 1);
            });
        });
        describe('on ArrowDown', () => {
            it('when nothing is selected should select first answer', () => {
                const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key: 'ArrowDown', code: 'ArrowDown' });
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[0]);
            });

            describe('when selection already present', () => {
                it('and not last', () => {
                    const selected = 1;
                    const shouldSelect = selected + 1;
                    const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                    fireEvent.click(getByText(container, allAnswers[selected]));
                    fireEvent.keyDown(container.firstChild!, { key: 'ArrowDown', code: 'ArrowDown' });
                    expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[shouldSelect]);
                });

                it('and is last', () => {
                    const selected = allAnswers.length - 1;
                    const shouldSelect = 0;
                    const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                    fireEvent.click(getByText(container, allAnswers[selected]));
                    fireEvent.keyDown(container.firstChild!, { key: 'ArrowDown', code: 'ArrowDown' });
                    expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[shouldSelect]);
                });
            });
            it('should not change selection in reveal mode', () => {
                const { container, getByTestId, queryAllByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.click(getByText(container, allAnswers[correctAnswerId]));
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[correctAnswerId]);
                fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' });
                fireEvent.keyDown(container.firstChild!, { key: 'ArrowDown', code: 'ArrowDown' });
                expect(getByTestId(AnswerStates.CorrectSelected)).toHaveTextContent(allAnswers[correctAnswerId]);
                expect(queryAllByTestId(AnswerStates.Notselected)).toHaveLength(allAnswers.length - 1);
            });
        });
        describe('on numeric keys', () => {
            it('when key is (0)', () => {
                const key = '0';
                const { container, queryAllByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key, code: key });
                expect(queryAllByTestId(AnswerStates.Selected)).toHaveLength(0);
            });
            it('when key is (1)', () => {
                const key = '1';
                const expectedSelectionId = +key - 1;
                const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key, code: key });
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[expectedSelectionId]);
            });
            it('when key is (2)', () => {
                const key = '2';
                const expectedSelectionId = +key - 1;
                const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key, code: key });
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[expectedSelectionId]);
            });
            it('when key is (3)', () => {
                const key = '3';
                const expectedSelectionId = +key - 1;
                const { container, getByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key, code: key });
                expect(getByTestId(AnswerStates.Selected)).toHaveTextContent(allAnswers[expectedSelectionId]);
            });
            it('when key is greater than num of answers (4)', () => {
                const key = '4';
                const { container, queryAllByTestId } = render(<Provider store={store}><Quiz /></Provider>);
                fireEvent.keyDown(container.firstChild!, { key, code: key });
                expect(queryAllByTestId(AnswerStates.Selected)).toHaveLength(0);
            });
        });
    });
    describe('should display summary with all questions and correct answers', () => {
        const { questions } = initialState.quiz;
        const { getByText, container } = render(<Provider store={store}><Quiz /></Provider>);
        questions.forEach(q => {
            fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' }); // reveal correct answer
            fireEvent.keyDown(container.firstChild!, { key: 'Enter', code: 'Enter' }); // load new question
        });
        expect(getByText(BUTTON_LABEL.RESULTS)).toBeTruthy();
        questions.forEach(q => {
            expect(getByText(q.question)).toBeTruthy();
            expect(getByText(q.correct_answer)).toBeTruthy();
        });
    });
});