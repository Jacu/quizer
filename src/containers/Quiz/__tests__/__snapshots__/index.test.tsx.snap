// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should render with props 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm ejebrS"
    tabindex="0"
  >
    <div
      class="sc-AxiKw cvapgi"
    >
      <p
        class="sc-AxgMl cqiLXF"
      >
        Category: mock
      </p>
      <h3
        class="sc-AxhUy yVvfC"
      >
        Is test passing?
      </h3>
      <div
        class="sc-AxheI fFzjlp"
      >
        <div
          class="sc-Axmtr kbPAJs"
          data-testid="NS"
        >
          <div
            class="sc-AxmLO bTFxKa"
          />
          no
        </div>
        <div
          class="sc-Axmtr kbPAJs"
          data-testid="NS"
        >
          <div
            class="sc-AxmLO bTFxKa"
          />
          of course!
        </div>
        <div
          class="sc-Axmtr kbPAJs"
          data-testid="NS"
        >
          <div
            class="sc-AxmLO bTFxKa"
          />
          not yet
        </div>
      </div>
    </div>
    <div
      class="sc-AxirZ fGKEvi"
    >
      <div
        class="sc-fznKkj etFNJk"
        size="1.25"
      >
        Check
      </div>
    </div>
    <div
      class="sc-fzoLsD ioAKZq"
    >
      <div
        class="sc-fzplWN fKnvyl"
      >
        Question 1 of 2
      </div>
      <div
        class="sc-fznyAO hrrMRl"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-home fa-w-18 sc-fzpans dsUZsr"
          data-icon="home"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
